### DISCUSSION TITLE
Appwrite Doc-Permissions

### DISCUSSION THREAD

**Discussion by EricKrg** [appwrite/appwrite#4781] [Q&A] [answered]
Hi guys,

i am still struggling a bit with the right and secure way off creating permissions in appwrite, i already read through https://github.com/appwrite/appwrite/discussions/2689 but didnt found what i was searching for. 

My Problem is the following, i have a collection with todo-items each user can create documents (todo-items) in that collection, with the creation of a document the following doc-permissions are set:
```js
database.createDocument(Server.databaseID, Server.collectionID,
                'unique()',
                JSON.stringify({ task: task, taskState: taskState }),
                [
                    Permission.update(Role.user(user.$id)),
                    Permission.delete(Role.user(user.$id)),
                    Permission.read(Role.user(user.$id)),
                    Permission.write(Role.user(user.$id)),
                ]);
```
That works fine and as expected. 
My Question is how can i prevent that a user is creating documents with permissions i dont want him/her to set? If a user knows how the appwrite rest-services works he/she can easily grab a valid token from the browser session and create documents with permission `Role.any()` for example. Reading through the discussions and documentation, setting the permissions of a doc. you own seems to be intended but i wonder if there is something i missed for my described use case.

**Comment by stnguyen90** [comment, marked as answer]
If you're really concerned about this, you can restrict create access and create an Appwrite Function to handle creating the documents. The Function can set the permission based on the user rather than letting the user put anything they want. 

Another option is to use an Appwrite function that triggers on document creation and delete the document if anything is invalid (like any for permission).

Fyi, you don't need to stringify the data. Also, you shouldn't pass write permission if you're already passing update and delete.

**Comment by EricKrg** [comment]
@stnguyen90 thank you for the quick response! I was thinking of using a function for this but wasn't sure if thats the appwrite-way of doing things.  

Thanks also for the pointer at the end!
